src/overlays/actors/ovl_Bg_Heavy_Block/z_bg_heavy_block.c

```diff
@@ -21,35 +38,497 @@ const ActorInit Bg_Heavy_Block_InitVars = {
     (ActorFunc)BgHeavyBlock_Update,
     (ActorFunc)BgHeavyBlock_Draw,
 };
-*/
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Heavy_Block/func_80883790.s")
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Heavy_Block/func_80883820.s")
+static InitChainEntry sInitChain[] = {
+    ICHAIN_VEC3F(scale, 1, ICHAIN_CONTINUE),
+    ICHAIN_F32(uncullZoneForward, 4000, ICHAIN_CONTINUE),
+    ICHAIN_F32(uncullZoneScale, 400, ICHAIN_CONTINUE),
+    ICHAIN_F32(uncullZoneDownward, 400, ICHAIN_STOP),
+};
+
+extern UNK_TYPE D_0600169C;
+extern Gfx D_060013C0[];
+extern Gfx D_06001A30[];
+extern Gfx D_060018A0[];
+
+void BgHeavyBlock_SetPieceRandRot(BgHeavyBlock* this, f32 scale) {
+    this->dyna.actor.posRot.rot.x = Math_Rand_CenteredFloat(1024.0f) * scale;
+    this->dyna.actor.posRot.rot.y = Math_Rand_CenteredFloat(1024.0f) * scale;
+    this->dyna.actor.posRot.rot.z = Math_Rand_CenteredFloat(1024.0f) * scale;
+}
+
+void BgHeavyBlock_InitPiece(BgHeavyBlock* this, f32 scale) {
+    f32 rand;
+    f32 yawSinCos;
+    f32 randChoice;
+
+    this->dyna.actor.gravity = -0.6f;
+    this->dyna.actor.minVelocityY = -12.0f;
+    randChoice = Math_Rand_CenteredFloat(12.0f * scale);
+    rand = (randChoice < 0.0f) ? randChoice - 2.0f : randChoice + 2.0f;
+    this->dyna.actor.velocity.y = (Math_Rand_ZeroFloat(8.0f) + 4.0f) * scale;
+    this->dyna.actor.velocity.z = Math_Rand_ZeroFloat(-8.0f * scale);
+    yawSinCos = Math_Coss(this->dyna.actor.posRot.rot.y);
+    this->dyna.actor.velocity.x =
+        (Math_Sins(this->dyna.actor.posRot.rot.y) * this->dyna.actor.velocity.z + (yawSinCos * rand));
+    yawSinCos = Math_Sins(this->dyna.actor.posRot.rot.y);
+    this->dyna.actor.velocity.z =
+        (Math_Coss(this->dyna.actor.posRot.rot.y) * this->dyna.actor.velocity.z) + (-yawSinCos * rand);
+    BgHeavyBlock_SetPieceRandRot(this, scale);
+    Actor_SetScale(&this->dyna.actor, Math_Rand_CenteredFloat(0.2f) + 1.0f);
+}
+
+void BgHeavyBlock_SetupDynapoly(BgHeavyBlock* this, GlobalContext* globalCtx) {
+    s32 pad[2];
+    UNK_TYPE a1;
+
+    a1 = 0;
+    this->dyna.actor.flags |= 0x20030;
+    DynaPolyInfo_SetActorMove(&this->dyna, 0);
+    DynaPolyInfo_Alloc(&D_0600169C, &a1);
+    this->dyna.dynaPolyId = DynaPolyInfo_RegisterActor(globalCtx, &globalCtx->colCtx.dyna, &this->dyna.actor, a1);
+}
+
+void BgHeavyBlock_Init(Actor* thisx, GlobalContext* globalCtx) {
+    BgHeavyBlock* this = THIS;
+
+    Actor_ProcessInitChain(thisx, sInitChain);
+    ActorShape_Init(&thisx->shape, 0.0f, NULL, 0.0f);
+    this->pieceFlags = 0;
+
+    if (globalCtx->sceneNum == SCENE_GANON_TOU) {
+        thisx->params &= 0xFF00;
+        thisx->params |= 4;
+    }
+
+    switch (thisx->params & 0xFF) {
+        case 2: // bigger broken piece
```

Comment:
All these case comments is where a "type" enum would make documentation much better and more consistent.
I'm not sure why making enums like that isn't already a natural part of the actor documentation process tbh.
And just to be clear, such enums should often go in the header, when they are intended for external use too (eg. when spawning actors dynamically or when defining spawn lists).

Reply:
sometimes you say that its too early for this type of stuff and other times you say that its not, so im never sure lmao. But id be glad to

Reply:
Making actor specific type enums is something I did on actors I decompiled literally a week into the project, so I doubt I ever said anything about this kind of stuff being too early.

In fact I have said a few times that I think making proper file-scopped enums/defines is just as important if not more important than comments or function-level documentation.

Reply:
Maybe Im confusing params bit mask macros with enums naming them. They were one in the same to me, but if we can do one butnot the other id like to clarify that now

Reply:
It's a completely different story for enums/defines that affect the whole codebase tho, since we can't really be sure of the full scope yet.

Reply:
I see, if thats where we draw the line then im cool with that. Ill keep it in mind going forward, thank you

Reply:
Macros are also somewhat another story because they depend directly on codegen and should be designed with the whole project in mind, even if they are specific to 1 actor for example.

Reply:
naming type after its usage is probably not great practice, but in the case of the block outside of double defense, the position and rotation is hard coded in init. Should the name be something like HEAVYBLOCK_OUTSIDE_CASTLE in this case?

Reply:
adding "outside castle" is probably fine, but I would have it on top of the name the "normal" type or whatever, assuming that's a thing.

Reply:
yea theres HEAVYBLOCK_STAY so ill do HEAVYBLOCK_STAY_OUTSIDE_CASTLE

Reply:
Alright, but I'm not sure "stay" really makes sense as the name for a type of the actor

Reply:
well it either breaks on impact or stays where it lands. is there a better term for that?

Reply:
In that case names like "breakable" or "unbreakable" would probably be more appropriate, or any other wordings that "describe" the types themselves, and not just indicate what they do with a verb.

Reply:
yea that sounds good, will do